### PR TITLE
Double (and Float) to Decimal Conversion Precision Improvement

### LINKED ISSUE #72135: Double to decimal conversion loses precision
I'm making a separate issue to discuss what was discovered [here](https://github.com/dotnet/runtime/issues/68042#issuecomment-1100258233) in more detail.

## Summary
The current double to decimal conversion code relies on some false assumptions. Namely, it assumes a double can only represent 15 base-10 digits of precision, which is incorrect. This leads to quite a number of inaccuracies when converting. I've included some examples below. I also included the result of constructing the decimal via string parsing to show that it is entirely possible for a decimal to precisely represent each value.

# Reproduction
```
double doubleValue = [INSERT VALUE HERE];
Console.WriteLine("Expected Result:");
Console.WriteLine(doubleValue.ToString("G99"));

decimal decimalValue = (decimal)doubleValue;
Console.WriteLine("Constructing Decimal Via Casting:"); 
Console.WriteLine(decimalValue.ToString("G99"));


decimal decimalValueParsed = decimal.Parse(doubleValue.ToString("G99"), System.Globalization.NumberStyles.Float);
Console.WriteLine("Constructing Decimal Via Parsing:"); 
Console.WriteLine(decimalValueParsed.ToString("G99"));
```
Outputs:
| Expected Result | Constructing Decimal Via Casting |  Constructing Decimal Via Parsing | 
| ----------- | ----------- |  ----------- |
| 10000000000000.099609375      | 10000000000000.1 |  10000000000000.099609375 |
| 79228162514264328797450928128 | 79228162514264300000000000000 | 79228162514264328797450928128 |
| 1.229999999999999982236431605997495353221893310546875 | 1.23 | 1.229999999999999982236431606 |

# Fix
I have a PR coming up soon with a fix to this. At a high level, we can take advantage of sections of our existing double to string conversion algorithms, which are performing that base-2 to base-10 conversion at much better precision.

# Impact
This issue has been around for as long as I can tell, so it is hard to say how high of a priority this fix is. I did run into this very bug while using the [PowerToys calculator](https://github.com/microsoft/PowerToys/issues/18574), which makes me believe this fix has high priority. Not being able to trust the output of your calculator because the framework it is built on has conversion bugs is concerning.

However, because this behavior has existed for so long, it is important to recognize that users might have incorrect expectations that we will be breaking with this fix. For example, a test case I found was assuming the double literal `1.23` could be represented exactly and casting it to decimal only validated their false assumption. With the fix applied, the more precise decimal we now get is not what they were expecting.

```
using System;

double doubleValue = 1.23;
Console.WriteLine("Actually stored as:");
Console.WriteLine(doubleValue.ToString("G99")); // 1.229999999999999982236431605997495353221893310546875

decimal decimalValue = (decimal)doubleValue;

var builder = new System.Text.StringBuilder("Hello");
builder.Append(decimalValue);
Console.WriteLine(builder.ToString()); 
// This currently (incorrectly) prints "Hello1.23"
// With the fix it prints "Hello1.22299999999999999822364316059"
// User should have used decimal literal 1.23m
```

### PR DESCRIPTION
**Pull request by dakersnar**
Fixes https://github.com/dotnet/runtime/issues/72135, https://github.com/microsoft/PowerToys/issues/18574

Blocked by https://github.com/dotnet/runtime/pull/72115 (although that change is included in this PR too)

See https://github.com/dotnet/runtime/issues/72135 for an explanation of the problem

## Summary of fix
This PR is a general refactor of the conversion code. The primary goal is to solve the reported bug, but secondarily make this conversion more efficient by reusing fast code we have written elsewhere. To achieve this, @tannergooding and I brainstormed reusing a section of the Dragon4 algorithm for double to string conversion, as printing a double already involves converting it to base 10.

To do this, I split up the code for the Dragon4 algorithm into two parts and added an additional API that allows Decimal.DecCalc.cs access to the first half of Dragon4. This API is structured to return everything Decimal.DecCalc.cs needs to construct a decimal representation of the original double.

## Testing
To see some of the impact of the change, take a look at how much existing test data had to be updated in `DecimalTests.cs`. Most of the expected results were wrong but the tests were passing because (I assume) they were built to match the output of the incorrect converter. To ensure that I wasn't relying on my conversion code as the "source of truth" to test itself, I generated the expected results using the decimal parser.

## Problems/TODO
- https://github.com/dotnet/runtime/issues/72125 - Decimal to Double conversion (the reverse direction) has some slight precision loss issues, preventing me from including a round trip test.
- Often, a decimal can be represented multiple valid ways. For example, for a value of .1, a mantissa of 10 with a scale of 2 is the same as a mantissa of 100 with a scale of 3. To normalize the output of the conversion, I tried to mirror the behavior of `decimal.GetBits(decimal.Parse(doubleValue.ToString("G99"), System.Globalization.NumberStyles.Float)` by removing trailing zeros for non-zero values. However, in some cases, the parser ends up retaining some trailing zeros, and there is a mismatch. In these cases, the output of my converter is still correct, it is just not completely consistent with what the parser will generate.
- Notably, I haven't perf tested this new converter yet.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Number.Dragon4.cs [hunk 1]
```diff
@@ -537,5 +561,171 @@ private static unsafe uint Dragon4(ulong mantissa, int exponent, uint mantissaHi
             Debug.Assert(outputLen <= buffer.Length);
             return outputLen;
         }
+
+        private static unsafe Dragon4State Dragon4GetState(ulong mantissa, int exponent, uint mantissaHighBitIdx, bool hasUnequalMargins, int cutoffNumber)
```
Comment (dakersnar) [inline src/libraries/System.Private.CoreLib/src/System/Number.Dragon4.cs:565]: Note, none of this function is new, git is just confused by my refactor. This is just the first half of the existing Dragon4 code hoisted to a helper function.

File: src/libraries/System.Private.CoreLib/src/System/Number.BigInteger.cs [hunk 2]
```diff
@@ -602,6 +602,11 @@ public static void DivRem(ref BigInteger lhs, ref BigInteger rhs, out BigInteger
                         {
                             remLength--;
                         }
+                        else
+                        {
+                            // As soon as we find a non-zero block, the rest of remainder is significant
+                            break;
+                        }
```
Comment (dakersnar) [inline src/libraries/System.Private.CoreLib/src/System/Number.BigInteger.cs:609]: This change has its own separate PR here: https://github.com/dotnet/runtime/pull/72115

**Comment by danmoseley** [comment]
@dakersnar What is the status of this PR? If it's blocked or work is deferred we should change it to draft status.

**Comment by dakersnar** [comment]
We determined the change was a bit too drastic to merge that late in the release cycle for .NET 7.0. We are planning to merge it into .NET 8. Should I mark it as a draft for now?

**Comment by danmoseley** [comment]
@dakersnar up to you, but in general we mark PR's draft when they aren't close to mergeable, eg., work has paused, design discussion needs to happen, it's blocked, etc. I believe validation can still be triggered on draft PR's, but e.g., we no longer report it in the "stale PR's" mailers.

Another option when work on a PR is being deferred (not suggesting here) is to close it, as it's one click to reopen. The only side effect is that it reruns all validation when you do that, which is sometimes desirable sometimes not.

**Comment by dakersnar** [comment]
This should be pretty much ready to review but converting it to draft for now as we don't plan on doing that for a bit. Thanks for the suggestion.

**Pull request closed without merging**